src/GraphQL/TypeApi.hs

```diff
@@ -245,40 +250,47 @@ instance forall typeName interfaces fields m.
   buildResolver mHandler selectionSet = do
     -- First we run the actual handler function itself in IO.
     handler <- mHandler
-    -- We're evaluating an Object so we're collecting (name, Value)
-    -- pairs from runFields and build a GValue.Map with them.
-    r <- forM selectionSet $ \selection -> runFields @m @fields handler selection
-    pure $ GValue.toValue (GValue.mapFromList r)
+    -- We're evaluating an Object so we're collecting ObjectFields from
+    -- runFields and build a GValue.Map with them.
+    r <- forM selectionSet $ runFields @m @fields handler
+    case GValue.makeObject r of
+      Nothing -> queryError $ "Duplicate fields in set: " <> show r
+      Just object -> pure $ GValue.ValueObject object
 
 
 -- | Closed type family to enforce the invariant that Union types
 -- contain only Objects.
 type family RunUnionType m (a :: [Type]) :: Type where
-  RunUnionType m ((Object typeName interfaces fields):rest) = HandlerType m (Object typeName interfaces fields) :<|> RunUnionType m rest
+  RunUnionType m (Object typeName interfaces fields:rest) = HandlerType m (Object typeName interfaces fields) :<|> RunUnionType m rest
   RunUnionType m '[] = ()
   RunUnionType m a = TypeLits.TypeError ('TypeLits.Text "All types in a union must be Object. Got: " 'TypeLits.:<>: 'TypeLits.ShowType a)
 
 
 -- Type class to execute union type queries.
 class RunUnion m a where
-  runUnion :: RunUnionType m a -> AST.Selection -> m GValue.Value
+  runUnion :: RunUnionType m a -> AST.Selection -> m GValue.Object
 
 instance forall m typeName interfaces fields rest.
          ( RunUnion m rest
          , MonadIO m
          , MonadThrow m
          , RunFields m fields
          , KnownSymbol typeName
-         ) => RunUnion m ((Object typeName interfaces fields):rest) where
+         ) => RunUnion m (Object typeName interfaces fields:rest) where
   runUnion (lh :<|> rh) fragment@(AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType queryTypeName) [] subSelection))
-    | typeName == queryTypeName = buildResolver @m @(Object typeName interfaces fields) lh subSelection
+    | typeName == queryTypeName = do
+        result <- buildResolver @m @(Object typeName interfaces fields) lh subSelection
+        -- TODO: See if we can prevent this from happening at compile time.
```

Comment:
Well, I was thinking more that we could have a separate resolver typeclass for things that are guaranteed to resolve to an object. Then the current resolver typeclass would be a superclass of that (if we know how to interpret a thing as an object, we know how to interpret it as a value). 

Type-level sets would presumably be for enforcing uniqueness.